apps/sms/views/shared/js/navigation.js

```diff
@@ -1,361 +1,921 @@
+        'panel-will-activate', 'panel-animation-reverse'
+      );
     });
+  }
 
-    promise = promise.then(
-      function resolved() {
-        this.transitionPromise = null;
-        if (nextPanelContainer === currentPanelContainer) {
-          return;
-        }
-        nextPanelContainer.setAttribute('aria-hidden', false);
-        if (currentPanelContainer) {
-          currentPanelContainer.setAttribute('aria-hidden', true);
-        }
-      }.bind(this),
-      function rejected(e) {
-        catchError(e);
-        this.transitionPromise = null;
-        return Promise.reject(new Error('Error while transitioning'));
-      }.bind(this)
-    );
+  if (oldView) {
+    oldPanelElement.classList.remove('panel-active');
+  }
 
-    promise.
-      then(() => this.emit('navigated', { panel: panel, args: args })).
-      then(nextQueuedPanel, nextQueuedPanel);
+  if (newView) {
+    newPanelElement.classList.add('panel-active');
+  }
 
-    return (this.transitionPromise = promise);
-  },
+  return Promise.resolve();
+}
+
+/**
+ * Reacts to hash changes.
+ *
+ * This is the heart of the navigation logic and handles running the various
+ * navigation steps.
+ * It also handles any unwanted errors. (TODO: needs to tune this a little
+ * more).
+ */
+function onHashChange() {
+  debug('onHashChange, location=', window.location);
+  var newPanel = findViewFromLocation(window.location);
+  if (!newPanel) {
+    console.error('Couldn\'t find a view from location', window.location);
+    return;
+  }
 
+  var args = Utils.params(decodeURIComponent(window.location.hash));
+
+  startNavigation({ newPanel, args }).then(
+    () => executeNavigationStep('beforeLeave')
+  ).then(
+    prepareSwitchPanel
+  ).then(
+    () => executeNavigationStep('beforeEnter').catch(catchStepError)
+  ).then(
+    setFutureState
+  ).then(
+    switchPanel
+  ).then(
+    () => executeNavigationStep('afterEnter').catch(catchStepError)
+  ).then(
+    () => executeNavigationStep('afterLeave').catch(catchStepError)
+  ).then(
+    endNavigation,
+    onNavigationError // TODO reset to previous hash or call back ?
+  );
+}
+
+/**
+ * Attaches the hashchange event handler.
+ */
+function attachHistoryListener() {
+  window.addEventListener('hashchange', onHashChange);
+}
+
+/**
+ * Removes the hashchange event handler.
+ */
+function detachHistoryListener() {
+  window.removeEventListener('hashchange', onHashChange);
+}
+
+var Navigation = {
   /**
-   * Navigates to default panel using specified arguments.
-   * @param {*?} args Any arguments that will be passed into default panel
-   * lifecycle methods.
-   * @returns {Promise} Promise that will be resolved once navigation to default
-   * panel is completed.
+   * Calls `window.back` after running the `beforeLeave` step.
+   *
+   * @returns {Promise} Resolves when the action is done.
    */
-  toDefaultPanel: function n_toDefaultPanel(args) {
-    return this.toPanel(this.defaultPanel, args);
+  back() {
```

Comment:
__nit:__ maybe I've missed, but it would be great to mention in comment for the navigation methods ```back``` and ```toPanel``` that the rest of the the work will be handled in ```onHashChange``` - it's a bit hard to understand when you read code for the first time :)